### PR TITLE
rgw/sse-s3: fix bucket encryption of multipart upload

### PR DESCRIPTION
**Pull request by 5cs**
Multipart upload missing encryption when we have bucket encryption policy. Fix it by fetching bucket encryption policy and resolving defaults at multipart init op.

Fixes: https://tracker.ceph.com/issues/59218

Signed-off-by: Tongliang Deng <dengtongliang@gmail.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [x] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -3876,10 +3876,19 @@ void RGWSetRequestPayment_ObjStore_S3::send_response()
 
 int RGWInitMultipart_ObjStore_S3::get_params(optional_yield y)
 {
+  int ret;
+
+  map_qs_metadata(s, true);
```
Comment (mdw-at-linuxbox) [inline src/rgw/rgw_rest_s3.cc:3881]: Is this call necessary?  The aws doc doesn't appear to document using query parameters to set encryption parameters when creating a multipart upload.

**Review by mdw-at-linuxbox — COMMENTED** [review]
What tracker ID goes with this fix?  This should be in the commit message.

**Comment by cbodley** [comment]
> What tracker ID goes with this fix? This should be in the commit message.

i've opened https://tracker.ceph.com/issues/59218 to track backports for this. we'll also want an s3test case that exercises this

File: src/rgw/rgw_rest_s3.cc (hunk 1 above)
Comment (5cs) [reply src/rgw/rgw_rest_s3.cc]: Removed.

**Comment by 5cs** [comment]
> > What tracker ID goes with this fix? This should be in the commit message.
> 
> i've opened https://tracker.ceph.com/issues/59218 to track backports for this. we'll also want an s3test case that exercises this

Thanks very much, I've updated the commit message with it.

**Comment by cbodley** [comment]
jenkins test make check

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
passed qa against tests from https://github.com/ceph/s3-tests/pull/505:
run https://pulpito.ceph.com/cbodley-2023-04-01_01:23:10-rgw-wip-cbodley-testing-distro-default-smithi/
rerun https://pulpito.ceph.com/cbodley-2023-04-03_13:05:03-rgw-wip-cbodley-testing-distro-default-smithi/

> test_sse_s3_default_multipart_upload PASSED

**Pull request merged**